File: 算法/代码随想录算法PDF/00调试源代码/20.有效的括号.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
public:
    bool isValid(string s) {
        if (s.size() % 2) {  // 奇数必不满足，剪枝
            return false;
        }
        stack<char> st;
        for (char c : s) {
            if (c == '(') {
                st.push(')');
            } else if (c == '[') {
                st.push(']');
            } else if (c == '{') {
                st.push('}');
            } else if (st.empty() || st.top() != c) { // 半中间栈空了或右括号匹配不上
                return false;
            } else {
                st.pop();
            }
        }
        return st.empty();
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2:  # 奇数一定不满足，剪枝
            return False
        st = []
        mapping = {  # 对照表
            '(' : ')',
            '[' : ']',
            '{' : '}',
        }
        for c in s:
            if c in mapping.keys():
                st.append(mapping[c])
            elif not st or st[-1] != c:
                return False
            else:
                st.pop()
        return not st
# @lc code=end

